x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,391 @@
+	}
+
+	// approve upgrade by user with Trustee role
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress2)
+	_, err = setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.NoError(t, err)
+
+	setup.UpgradeKeeper.AssertCalled(
+		t,
+		"ScheduleUpgrade",
+		mock.MatchedBy(isContextWithNonCachedMultiStore),
+		msgProposeUpgrade.Plan,
+	)
+}
+
+func TestHandler_ProposedUpgradeDoesNotExist(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress2 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress3 := sdktestdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// approve upgrade
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress1)
+	_, err := setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.Error(t, err)
+	require.True(t, types.ErrProposedUpgradeDoesNotExist.Is(err))
+}
+
+func TestHandler_MessageCreatorAlreadyApprovedUpgrade(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress2 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress3 := sdktestdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// propose new upgrade
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+
+	// approve upgrade
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress1)
+	_, err = setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.Error(t, err)
+	require.True(t, sdkerrors.ErrUnauthorized.Is(err))
+}
+
+func TestHandler_PlanHeightLessBlockHeight(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := sdktestdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(1)
+
+	// propose new upgrade
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+	msgProposeUpgrade.Plan.Height = 1
+	setup.Ctx = setup.Ctx.WithBlockHeight(100)
+	blockHeight := setup.Ctx.BlockHeight()
+
+	require.Less(t, msgProposeUpgrade.Plan.Height, blockHeight)
+
+	setup.ScheduleUpgradeBehavior(msgProposeUpgrade.Plan, sdkerrors.ErrInvalidRequest)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
```

Comment:
We need also a similar test that faced a too low height on approval step (where the earlier proposal was successful).

Reply:
Added